File: backend/LiveSync.AuthApi/Services/AuthService.cs
using Microsoft.AspNetCore.Identity;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using LiveSync.AuthApi.DTOs;
using LiveSync.AuthApi.Models;

namespace LiveSync.AuthApi.Services
{
    public class AuthService : IAuthService
    {
        private readonly UserManager<ApplicationUser> _userManager;
        private readonly SignInManager<ApplicationUser> _signInManager;
        private readonly IConfiguration _configuration;
        private readonly ILogger<AuthService> _logger;

        public AuthService(
            UserManager<ApplicationUser> userManager,
            SignInManager<ApplicationUser> signInManager,
            IConfiguration configuration,
            ILogger<AuthService> logger)
        {
            _userManager = userManager;
            _signInManager = signInManager;
            _configuration = configuration;
            _logger = logger;
        }

        public async Task<AuthResponse> RegisterAsync(RegisterRequest request)
        {
            try
            {
                // Check if user already exists
                var existingUser = await _userManager.FindByEmailAsync(request.Email);
                if (existingUser != null)
                {
                    return new AuthResponse
                    {
                        Success = false,
                        Message = "User with this email already exists."
                    };
                }

                // Create new user
                var user = new ApplicationUser
                {
                    UserName = request.Email,
                    Email = request.Email,
                    FirstName = request.FirstName,
                    LastName = request.LastName,
                    CreatedAt = DateTime.UtcNow
                };

                var result = await _userManager.CreateAsync(user, request.Password);

                if (!result.Succeeded)
                {
                    var errors = string.Join(", ", result.Errors.Select(e => e.Description));
                    return new AuthResponse
                    {
                        Success = false,
                        Message = $"Registration failed: {errors}"
                    };
                }

                // Generate JWT token
                var token = await GenerateJwtToken(user);

                return new AuthResponse
                {
                    Success = true,
                    Message = "Registration successful.",
                    Token = token,
                    Expiration = DateTime.UtcNow.AddHours(GetTokenExpirationHours()),
                    User = new UserInfo
                    {
                        Id = user.Id,
                        Email = user.Email!,
                        UserName = user.UserName,
                        FirstName = user.FirstName,
                        LastName = user.LastName
                    }
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error during registration");
                return new AuthResponse
                {
                    Success = false,
                    Message = "An error occurred during registration."
                };
            }
        }

        public async Task<AuthResponse> LoginAsync(LoginRequest request)
        {
            try
            {
                // Find user by email or username
                var user = await _userManager.FindByEmailAsync(request.EmailOrUsername)
                    ?? await _userManager.FindByNameAsync(request.EmailOrUsername);

                if (user == null)
                {
                    return new AuthResponse
                    {
                        Success = false,
                        Message = "Invalid credentials."
                    };
                }

                // Check password
                var result = await _signInManager.CheckPasswordSignInAsync(user, request.Password, lockoutOnFailure: true);

                if (!result.Succeeded)
                {
                    if (result.IsLockedOut)
                    {
                        return new AuthResponse
                        {
                            Success = false,
                            Message = "Account is locked out. Please try again later."
                        };
                    }

                    return new AuthResponse
                    {
                        Success = false,
                        Message = "Invalid credentials."
                    };
                }

                // Update last login time
                user.LastLoginAt = DateTime.UtcNow;
                await _userManager.UpdateAsync(user);

                // Generate JWT token
                var token = await GenerateJwtToken(user);

                return new AuthResponse
                {
                    Success = true,
                    Message = "Login successful.",
                    Token = token,
                    Expiration = DateTime.UtcNow.AddHours(GetTokenExpirationHours()),
                    User = new UserInfo
                    {
                        Id = user.Id,
                        Email = user.Email!,
                        UserName = user.UserName,
                        FirstName = user.FirstName,
                        LastName = user.LastName
                    }
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error during login");
                return new AuthResponse
                {
                    Success = false,
                    Message = "An error occurred during login."
                };
            }
        }

        public async Task<AuthResponse> RefreshTokenAsync(string token)
        {
            // TODO: Implement refresh token logic
            // This would require storing refresh tokens in the database
            await Task.CompletedTask;
            
            return new AuthResponse
            {
                Success = false,
                Message = "Refresh token functionality not yet implemented."
            };
        }

        #region OAuth Methods - Placeholder for Future Implementation

        public async Task<AuthResponse> GoogleLoginAsync(OAuthLoginRequest request)
        {
            // TODO: Implement Google OAuth login
            // 1. Validate the access token with Google
            // 2. Get user info from Google
            // 3. Find or create user in database
            // 4. Generate JWT token
            
            await Task.CompletedTask;
            
            return new AuthResponse
            {
                Success = false,
                Message = "Google OAuth login not yet implemented."
            };
        }

        public async Task<AuthResponse> GitHubLoginAsync(OAuthLoginRequest request)
        {
            // TODO: Implement GitHub OAuth login
            // 1. Validate the access token with GitHub
            // 2. Get user info from GitHub
            // 3. Find or create user in database
            // 4. Generate JWT token
            
            await Task.CompletedTask;
            
            return new AuthResponse
            {
                Success = false,
                Message = "GitHub OAuth login not yet implemented."
            };
        }

        public async Task<AuthResponse> MicrosoftLoginAsync(OAuthLoginRequest request)
        {
            // TODO: Implement Microsoft OAuth login
            // 1. Validate the access token with Microsoft
            // 2. Get user info from Microsoft
            // 3. Find or create user in database
            // 4. Generate JWT token
            
            await Task.CompletedTask;
            
            return new AuthResponse
            {
                Success = false,
                Message = "Microsoft OAuth login not yet implemented."
            };
        }

        #endregion

        #region Private Helper Methods

        private async Task<string> GenerateJwtToken(ApplicationUser user)
        {
            var userRoles = await _userManager.GetRolesAsync(user);

            var claims = new List<Claim>
            {
                new Claim(ClaimTypes.NameIdentifier, user.Id),
                new Claim(ClaimTypes.Name, user.UserName ?? string.Empty),
                new Claim(ClaimTypes.Email, user.Email ?? string.Empty),
                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString())
            };

            // Add role claims
            foreach (var role in userRoles)
            {
                claims.Add(new Claim(ClaimTypes.Role, role));
            }

            var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(
                _configuration["Jwt:Secret"] ?? throw new InvalidOperationException("JWT Secret not configured")));
            var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);

            var token = new JwtSecurityToken(
                issuer: _configuration["Jwt:Issuer"],
                audience: _configuration["Jwt:Audience"],
                claims: claims,
                expires: DateTime.UtcNow.AddHours(GetTokenExpirationHours()),
                signingCredentials: creds
            );

            return new JwtSecurityTokenHandler().WriteToken(token);
        }

        private int GetTokenExpirationHours()
        {
            return _configuration.GetValue<int>("Jwt:ExpirationHours", 24);
        }

        #endregion
    }
}

File: livesync-api/services/auth_service.go
package services

import (
	"context"
	"errors"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/livesync/livesync-api/database"
	"github.com/livesync/livesync-api/models"
	"github.com/livesync/livesync-api/security"
)

type AuthService struct {
	db       *database.DB
	jwt      *security.JWTService
	password *security.PasswordHasher
}

func NewAuthService(db *database.DB, jwt *security.JWTService, password *security.PasswordHasher) *AuthService {
	return &AuthService{
		db:       db,
		jwt:      jwt,
		password: password,
	}
}

func (s *AuthService) Register(ctx context.Context, req *models.RegisterRequest) (*models.AuthResponse, error) {
	if req.Password != req.ConfirmPassword {
		return &models.AuthResponse{Success: false, Message: "Passwords do not match."}, nil
	}

	if err := security.ValidatePasswordStrength(req.Password); err != nil {
		return &models.AuthResponse{Success: false, Message: err.Error()}, nil
	}

	email := strings.TrimSpace(req.Email)
	normalizedEmail := strings.ToUpper(email)

	// Check if user already exists
	var exists bool
	checkQuery := `SELECT EXISTS(SELECT 1 FROM "AspNetUsers" WHERE "NormalizedEmail" = $1);`
	err := s.db.Pool.QueryRow(ctx, checkQuery, normalizedEmail).Scan(&exists)
	if err != nil {
		return nil, err
	}
	if exists {
		return &models.AuthResponse{Success: false, Message: "User with this email already exists."}, nil
	}

	preferredUserName := email
	if strings.TrimSpace(req.FirstName) != "" {
		preferredUserName = strings.TrimSpace(req.FirstName)
	}
	normalizedUserName := strings.ToUpper(preferredUserName)

	hashedPassword, err := s.password.Hash(req.Password)
	if err != nil {
		return nil, err
	}

	user := &models.ApplicationUser{
		ID:                   uuid.New().String(),
		Email:                &email,
		UserName:             &preferredUserName,
		PasswordHash:         &hashedPassword,
		FirstName:            &req.FirstName,
		LastName:             &req.LastName,
		CreatedAt:            time.Now(),
		NormalizedEmail:      &normalizedEmail,
		NormalizedUserName:   &normalizedUserName,
		EmailConfirmed:       false,
		SecurityStamp:        ptr(uuid.New().String()),
		ConcurrencyStamp:     ptr(uuid.New().String()),
		PhoneNumberConfirmed: false,
		TwoFactorEnabled:     false,
		LockoutEnabled:       true,
		AccessFailedCount:    0,
	}

	insertQuery := `
		INSERT INTO "AspNetUsers" (
			"Id", "Email", "UserName", "PasswordHash", "FirstName", "LastName",
			"CreatedAt", "NormalizedEmail", "NormalizedUserName", "EmailConfirmed",
			"SecurityStamp", "ConcurrencyStamp", "PhoneNumberConfirmed", "TwoFactorEnabled",
			"LockoutEnabled", "AccessFailedCount"
		) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16);
	`

	_, err = s.db.Pool.Exec(ctx, insertQuery,
		user.ID, user.Email, user.UserName, user.PasswordHash, user.FirstName, user.LastName,
		user.CreatedAt, user.NormalizedEmail, user.NormalizedUserName, user.EmailConfirmed,
		user.SecurityStamp, user.ConcurrencyStamp, user.PhoneNumberConfirmed, user.TwoFactorEnabled,
		user.LockoutEnabled, user.AccessFailedCount,
	)
	if err != nil {
		return nil, err
	}

	token, exp, err := s.jwt.Generate(user)
	if err != nil {
		return nil, err
	}

	return &models.AuthResponse{
		Success:    true,
		Message:    "Registration successful.",
		Token:      &token,
		Expiration: &exp,
		User: &models.UserInfo{
			ID:        user.ID,
			Email:     user.Email,
			UserName:  user.UserName,
			FirstName: user.FirstName,
			LastName:  user.LastName,
		},
	}, nil
}

func (s *AuthService) Login(ctx context.Context, req *models.LoginRequest) (*models.AuthResponse, error) {
	key := strings.ToUpper(strings.TrimSpace(req.EmailOrUsername))

	query := `
		SELECT "Id", "Email", "UserName", "PasswordHash", "FirstName", "LastName",
		       "CreatedAt", "LastLoginAt", "NormalizedEmail", "NormalizedUserName",
		       "EmailConfirmed", "SecurityStamp", "ConcurrencyStamp", "PhoneNumberConfirmed",
		       "TwoFactorEnabled", "LockoutEnabled", "AccessFailedCount", "LockoutEnd"
		FROM "AspNetUsers"
		WHERE "NormalizedEmail" = $1 OR "NormalizedUserName" = $1
		LIMIT 1;
	`

	var user models.ApplicationUser
	err := s.db.Pool.QueryRow(ctx, query, key).Scan(
		&user.ID, &user.Email, &user.UserName, &user.PasswordHash, &user.FirstName, &user.LastName,
		&user.CreatedAt, &user.LastLoginAt, &user.NormalizedEmail, &user.NormalizedUserName,
		&user.EmailConfirmed, &user.SecurityStamp, &user.ConcurrencyStamp, &user.PhoneNumberConfirmed,
		&user.TwoFactorEnabled, &user.LockoutEnabled, &user.AccessFailedCount, &user.LockoutEnd,
	)

	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return &models.AuthResponse{Success: false, Message: "Invalid credentials."}, nil
		}
		return nil, err
	}

	now := time.Now()
	if user.LockoutEnd != nil {
		if user.LockoutEnd.After(now) {
			return &models.AuthResponse{Success: false, Message: "Account is locked out. Please try again later."}, nil
		} else {
			user.LockoutEnd = nil
			user.AccessFailedCount = 0
		}
	}

	pwHash := ""
	if user.PasswordHash != nil {
		pwHash = *user.PasswordHash
	}

	if !s.password.Matches(req.Password, pwHash) {
		if user.LockoutEnabled {
			user.AccessFailedCount++
			if user.AccessFailedCount >= 5 {
				lockout := now.Add(5 * time.Minute)
				user.LockoutEnd = &lockout
			}
			_, _ = s.db.Pool.Exec(ctx,
				`UPDATE "AspNetUsers" SET "AccessFailedCount" = $1, "LockoutEnd" = $2 WHERE "Id" = $3;`,
				user.AccessFailedCount, user.LockoutEnd, user.ID,
			)
		}
		return &models.AuthResponse{Success: false, Message: "Invalid credentials."}, nil
	}

	// Login successful - reset lockout and update LastLoginAt
	user.AccessFailedCount = 0
	user.LockoutEnd = nil
	user.LastLoginAt = &now

	_, _ = s.db.Pool.Exec(ctx,
		`UPDATE "AspNetUsers" SET "AccessFailedCount" = 0, "LockoutEnd" = NULL, "LastLoginAt" = $1 WHERE "Id" = $2;`,
		now, user.ID,
	)

	token, exp, err := s.jwt.Generate(&user)
	if err != nil {
		return nil, err
	}

	return &models.AuthResponse{
		Success:    true,
		Message:    "Login successful.",
		Token:      &token,
		Expiration: &exp,
		User: &models.UserInfo{
			ID:        user.ID,
			Email:     user.Email,
			UserName:  user.UserName,
			FirstName: user.FirstName,
			LastName:  user.LastName,
		},
	}, nil
}

func ptr[T any](v T) *T {
	return &v
}
